go/ql/lib/semmle/go/frameworks/Twirp.qll

```diff
@@ -0,0 +1,175 @@
+
+    /**
+     * Returns the named type on top of this interface type
+     */
+    NamedType getNamedType() { result = serviceInterface }
+  }
+
+  /**
+   * A Twirp client
+   */
+  class ServiceClient extends NamedType {
+    ServiceClient() {
+      exists(ServiceInterface i, PointerType p |
+        p.implements(i) and
+        this = p.getBaseType() and
+        this.getName().toLowerCase() = i.getName().toLowerCase() + ["protobuf", "json"] + "client"
+      )
+    }
+  }
+
+  /**
+   * A Twirp server
+   */
+  class ServiceServer extends NamedType {
+    ServiceServer() {
+      exists(ServiceInterface i |
+        this.implements(i) and
+        this.getName().toLowerCase() = i.getName().toLowerCase() + "server"
+      )
+    }
+  }
+
+  /**
+   * A Twirp function to construct a Client
+   */
+  class ClientConstructor extends Function {
+    ClientConstructor() {
+      exists(ServiceClient c |
+        this.getName().toLowerCase() = "new" + c.getName().toLowerCase() and
+        this.getParameter(0).getType().getName() = "string" and
+        this.getParameter(1).getType().getName() = "HTTPClient"
+      )
+    }
+  }
+
+  /**
+   * A Twirp function to construct a Server
+   * Its first argument should be an implementation of the service interface
+   */
+  class ServerConstructor extends Function {
+    ServerConstructor() {
+      exists(ServiceServer c, ServiceInterface i |
+        this.getName().toLowerCase() = "new" + c.getName().toLowerCase() and
+        this.getParameter(0).getType() = i.getNamedType()
+      )
+    }
+  }
+
+  /**
+   * An SSRF sink for the Client constructor
+   */
+  class ClientRequestUrlAsSink extends RequestForgery::Sink {
+    ClientRequestUrlAsSink() {
+      exists(DataFlow::CallNode call |
+        call.getArgument(0) = this and
+        call.getTarget() instanceof ClientConstructor
+      )
+    }
+
+    override DataFlow::Node getARequest() { none() }
+
+    override string getKind() { result = "URL" }
+  }
+
+  /**
+   * A service handler
+   */
+  class ServiceHandler extends Method {
+    ServiceHandler() {
+      exists(DataFlow::CallNode call, Type handlerType, ServiceInterface i |
+        call.getTarget() instanceof ServerConstructor and
+        call.getArgument(0).getType() = handlerType and
+        handlerType.implements(i) and
+        this = handlerType.getMethod(_) and
+        this.implements(i.getNamedType().getMethod(_))
+      )
+    }
+  }
+
+  /**
+   * A request coming to the service handler
+   */
+  class Request extends UntrustedFlowSource::Range, DataFlow::ParameterNode {
+    ServiceHandler handler;
+
+    Request() {
+      handler.getParameter(0).getType().hasQualifiedName("context", "Context") and
+      handler.getParameter(_) = this.asParameter() and
+      this.getType().(PointerType).getBaseType() instanceof ProtobufMessage
+    }
+
+    override predicate isParameterOf(Callable c, int i) {
+      c.asFunction() = handler and
+      i = 1
+    }
+  }
+}
```

Comment:
I haven't seen anything modelled like this before. You're specifying a parameter of a method. It's much more common to specify an argument to a call to a method. Would that be suitable for this situation?

Reply:
I think its very common that, for web frameworks, untrusted data enters the application as the parameter of a handler/controller method. I can make it read accesses to that parameter but I dont think it makes sense to make it an argument to a call 🤔